### ISSUE TITLE
mangled URLs on entailment

### ISSUE THREAD

**Issue by doriantaylor** [ruby-rdf/rdf-reasoner#7]
I am not sure if this is a bug here or `RDF::Vocabulary`, but:

I have a routine that finds all subclasses/equivalent classes for a given class (it just recursively calls `subClass` and `equivalentClass` on whatever you feed it) and discovered it had stopped working on my own vocabulary. Turns out:
```
[4] pry(main)> RDF::SAK::CI.Audience.subClassOf
=> [#<RDF::URI:0x16d0 URI:dc:AgentClass>, #<RDF::URI:0x16e4 URI:skos:Concept>]
```
That's obviously wrong, ie it's not resolving the prefixes, but I hadn't changed anything. So I tried:
```
[5] pry(main)> require 'rdf/vocab/foaf'
=> true
[6] pry(main)> RDF::Vocab::FOAF.Image.subClassOf
=> [#<RDF::URI:0x1cfc URI:foaf:Document>]
```
…and it turns out it doesn't resolve in these either.

**Comment by gkellogg** [comment]
Well, something obviously broke, which wasn't picked up in the tests.

There was some change to how prefixes are found, due to deferred loading of vocabularies, and prefixes now depend on those specific vocabularies registered in the rdf and rdf-vocab libraries, so finding PNames for arbitrary vocabularies is impacted, but that shouldn't be the case here.

This will be an issue with the rdf gem, and not the reasoner.

Also, finding all subClasses (maybe equivlentClasses) should be directly supported by the reasoner using subClass entailment.

**Comment by gkellogg** [comment]
> That's obviously wrong, ie it's not resolving the prefixes, but I hadn't changed anything. So I tried:
> 
> ```
> [5] pry(main)> require 'rdf/vocab/foaf'
> => true
> [6] pry(main)> RDF::Vocab::FOAF.Image.subClassOf
> => [#<RDF::URI:0x1cfc URI:foaf:Document>]
> ```
> 
> …and it turns out it doesn't resolve in these either.

This is because you're not including rdf/vocab, which is where vocabularies are linked in. Try the following:

```ruby
require 'rdf/vocab'
RDF::Vocab::FOAF.Image.subClassOf
```

The previous version didn't rely on the extensions:

https://github.com/ruby-rdf/rdf-vocab/blob/7bb82695cd86b1ce93a0d045d3f467ca27ea6369/lib/rdf/vocab/extensions.rb#L36-L48

**Comment by gkellogg** [comment]
> Would it make sense to have something like `RDF::Vocab.register(prefix: 'path/to/vocab')`?
> 
> incidentally the code that puts `equivalentClass` and `subClass` together looks like this:
> 
> ```ruby
>   def all_related rdftype
>     t = RDF::Vocabulary.find_term(rdftype) or raise "No type #{rdftype.to_s}"
>     q = [t] # queue
>     c = {}  # cache
> 
>     while term = q.shift
>       # add term to cache
>       c[term] = term
> 
>       # keep this from tripping up
>       next unless term.uri? and term.respond_to? :class?
> 
>       # entail equivalent classes
>       term.entail(:equivalentClass).each do |ec|
>         # add equivalent classes to queue (if not already cached)
>         q.push ec unless c[ec]
>         c[ec] = ec unless ec == term
>       end
> 
>       # entail subclasses
>       term.subClass.each do |sc|
>         # add subclasses to queue (if not already cached)
>         q.push sc unless c[sc]
>         c[sc] = sc unless sc == term
>       end
>     end
> 
>     # smush the result 
>     c.keys
>   end
> ```

You could try the following:

```ruby
require 'rdf/reasoner'
require 'rdf/vocab'

RDF::Reasoner.apply(:rdfs, :owl)

RDF::Vocab::FOAF.Person.entail(:subClassOf)
RDF::Vocab::SCHEMA.DataCatalog.entail(:equivalentClass)
```

**Comment by gkellogg** [comment]
@doriantaylor is this issue resolved now?

**Comment by doriantaylor** [comment]
I just updated everything and tried the test case and apparently not:

```
[1] pry(main)> require 'rdf/vocab/foaf'
=> true
[2] pry(main)> RDF::Vocab::FOAF.Image.subClassOf
=> [#<RDF::URI:0x15e0 URI:foaf:Document>]
[3] pry(main)> require 'rdf/reasoner'
=> true
[4] pry(main)> RDF::Reasoner.apply :rdfs, :owl
=> [:rdfs, :owl]
[5] pry(main)> RDF::Vocab::FOAF.Image.subClassOf
=> [#<RDF::URI:0x66440 URI:foaf:Document>]
```

**Comment by doriantaylor** [comment]
oh, no, wait a moment:

```
[1] pry(main)> require 'rdf'
=> true
[2] pry(main)> require 'rdf/vocab'
=> true
[3] pry(main)> RDF::Vocab::FOAF.Image.subClassOf
=> [#<RDF::Vocabulary::Term:0x21c ID:http://xmlns.com/foaf/0.1/Document>]
```

**Comment by doriantaylor** [comment]
OK it's coming back to me: I think the problem initially was that vocabs that weren't indexed in `RDF::Vocab` would not have their relations resolved. At the time, there was also a problem loading `RDF::Vocab`; it was really _really_ slow because it was `require`-ing stuff over and over again. I wrote `RDF::Vocabulary.register` so arbitrary vocabs could be added to the index.

(Speaking of which I guess I could add that `register` business to the vocab generation boilerplate.)

**Comment by gkellogg** [comment]
I believe when you start to reason, the subClass and other caches get built, so if you've evaluated something before requiring a vocabulary, it may not recalculate. Obviously, something could be added to clear caches, or it could be done as a consequence of including.

One thing we wanted with `register` was to be able to have finer control of what vocabularies were reasoned over, so we wouldn't want to remove this kind of fine-tuning.

Please continue on this, if you can.

**Comment by doriantaylor** [comment]
It looks like this resolution happens when `RDF::Vocab` is loaded, rather than anything to do with `RDF::Reasoner`. My understanding is vocabularies get generated with all applicable identifiers in CURIE form and those get resolved against a central mapping of prefixes by `RDF::Vocab`. I'm assuming `RDF::Reasoner` just generates caches out of assertions (of which e.g. `.subClassOf` is one, though its inverse, `.subClass`, is inferred). 

TL;DR this behaviour

**Comment by gkellogg** [comment]
The idea was to use PNames, where possible, in the vocabularies to reduce size and make more humanly comprehensible. If using the vocabulary and these PName prefixes aren’t resolved, that could cause an issue. Maybe they should require the vocabularies whose prefixes they depend on.

**Comment by gkellogg** [comment]
@doriantaylor please see if the latest updates address the issue.

**Comment by doriantaylor** [comment]
I keep circling back to this and then getting distracted. Mainly I'm thinking the larger question of prefixes in general might be worth reconsidering. The nub of the issue is that users of the library may have different preferences for the prefix of a given vocab, and `RDF::Vocab` appears to be opinionated about them, additionally when serializing vocabs. I have noticed elsewhere (eg in transforming/serializing graphs with the `rdf` tool) that the prefixes that ship with `RDF::Vocab` override the ones expressed in the input.

The issue with `RDF::Reasoner` not appropriately resolving pnames stems from the situation when a vocabulary is loaded without first loading `RDF::Vocab` and the prefix map, so my inclination would be to embed strictly local prefix mappings into `rdf-vocab/lib/rdf/vocab/*.rb` and have them inflate when they are loaded. That would save space and typing (assuming they weren't already being generated) and would be indifferent to whatever was going on in `RDF::Vocab`.

As for prefix mappings proper, I have found in my travels the need for two prefix maps: one for the purpose of abridging symbols in imperative code (a role ultimately played by `RDF::Vocab`) and another for transformation/serialization, and in principle they never see each other.

I'm not sure what the next steps are but I think this is an issue with `RDF::Vocab` and others, rather than `RDF::Reasoner`.

**Comment by gkellogg** [comment]
Writers should all take a `prefixes` option and use that for shortening URais before looking at vocabulary definitions, although this code is not centralized in RDF::Writer. Also, this likely doesn’t have tests.

the rdf Calai should take prefixes defined in a reader and passes them to the writer when serializing, to favor the original prefixes. But this is Ted used in `URI#pname`, IIRC.

if you’re seeing different behavior, then something has gone off somewhere.

does the pname resolution problem persist after the recent changes to largely eliminate them in Vocab? My original thought was to require dependent vocabularies in rdf-vocab, but instead changed to avoid the use of pnames altogether. 

probably, we need to create a couple of issues in rdf and rdf-vocab to track, ideally with test cases.

**Comment by gkellogg** [comment]
@doriantaylor I believe the problem where prefixes that are read are not re-used when writing in the CLI (and distiller) is fixed here: https://github.com/ruby-rdf/rdf/commit/0fb0185e8ca1f95ebe7a803cef9fe1ca19fa1ed4.

This would have been a casualty of the change to keyword parameters some time ago.

Generally, creating PNames from URIs happens when serializing, and by passing the `prefixes` found when reading to the writer, this should get most issues. Others occured when PNames were used internally, as with vocabulary definitions, but that was changed in RDF.rb 3.1.11 (and rdf-vocab 3.1.11). It _can_ still lookup PNames, but that's mostly for compatibility with old vocabularies, although it will expand and lookup using absolute URIs as well.
